Support kwargs in create child algorithm

gemmaguest opened the pull request:
Originally this was PR #32134 - see comments on that PR

**Description of work.**

This PR adds support for keyword args in createChildAlgorithm to reduce the need to write boilerplate code when calling child algorithms from python workflow algorithms. This was identified as a requirement as part of the work for #32125.

This changes some core functionality so please review carefully.

Similar code in IFunctionAdapter has been consolidated to use the same supporting class.

**To test:**

Try using createChildAlgorithm from python, e.g.:
```
# Create a child algorithm with a mixture of positional and keyword args
parent_alg = AlgorithmManager.createUnmanaged('Load')
args={"XUnit": "Wavelength", "XMax": 10000.0}
child_alg = parent_alg.createChildAlgorithm('CreateSampleWorkspace', 0.0, version=1, NumBanks=4, BinWidth=100.0, **args)
child_alg.execute()
ws=child_alg.getProperty("OutputWorkspace").value
AnalysisDataService.addOrReplace("ws", ws)

# Check we get the same results running the algorithm normally
ws2=CreateSampleWorkspace(**args, NumBanks=4, BinWidth=100)
CompareWorkspaces(Workspace1='ws', Workspace2='ws2', Tolerance=1e-08, CheckSample=True, ToleranceRelErr=True, CheckAllData=True)
```

Also check the history of a workflow algorithm that uses createChildAlgorithm, e.g.
- Run `ReflectometryISISLoadAndProcess('INTER13460', 0.5)`
- Show the history of workspace `IvsQ_binned_13460`
- Tick Unroll on the top level algorithm and then click `Script to Clipboard`
- Check that the results match the following (apart from the path and the temporary workspace name):
```
LoadNexus(Filename='/path/to/INTER00013460.nxs', OutputWorkspace='TOF_INTER13460')
RenameWorkspace(InputWorkspace='TOF_INTER13460', OutputWorkspace='TOF_13460')
ReflectometryReductionOneAuto(InputWorkspace='TOF_13460', ThetaIn=0.5, MomentumTransferMin=0.0064527921283052733, MomentumTransferStep=0.047640977777016841, MomentumTransferMax=0.072643025200851155, ScaleFactor=1, OutputWorkspaceBinned='IvsQ_binned_13460', OutputWorkspace='IvsQ_13460', OutputWorkspaceWavelength='__TMP0x7fdee203f400')
GroupWorkspaces(InputWorkspaces='TOF_13460', OutputWorkspace='TOF')
```

There are no workflow algorithms that currently use the new kwargs functionality so for now the main thing is to ensure the current functionality is not broken. A subsequent PR as part of #32125 will exercise the kwargs functionality.


*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Add overload to createChildAlgorithm with properties

gemmaguest committed:
Set properties by correct type

gemmaguest committed:
Add unit tests for kwargs of different types

gemmaguest committed:
Explicitly check python types for setting properties

Previously we were allowing the type checking to be done implicitly
which would mark int as bool etc.

gemmaguest committed:
Move python types extractor into core converters

gemmaguest committed:
Add support for property lists

Uses recursive variant so supports lists of lists etc.

gemmaguest committed:
Add unit test for list input properties

DavidFair committed:
Manually dispatch for std::vector<T> types

Boost will try to recurse down to the final level of a scalar value. In
our case we need to manually take over and dispatch for container types
using RTTI.

DavidFair committed:
Handle long/int mismatch between scalar and array props

IntArray is actually an array of longs, since we already have a long
type from Python we can preserve this until we try to set a scalar. In
the latter case we need to cast to an int for the underlying assignment
to work.

DavidFair committed:
Cleanup redundant assignments in NativeTypeExtractor

The compiler is smart enough to assign these types directly, this was a
workaround that was for the compiler not fully coercing the implicit
convertors for extractor types

DavidFair committed:
Move to generic impl to own class for dynamic dispatch

Move the dynamic dispatcher into it's own class. This makes it trivial
to use this mechanism to dynamically sort user inputs and handle them in
various different ways

DavidFair committed:
Switch IFunctionAdaptor to use common extractor

Ports IFunctionAdaptor to use the same adaptor whilst preserving the
same input limitations.

DavidFair committed:
Improve error handling and backwards compatibility

gemmaguest committed:
Avoid passing lists of mixed value types to algorithms

gemmaguest committed:
Add release notes with new createChildAlgorithm functionality

gemmaguest committed:
Minor tidying to avoid copy and reduce verbosity

gemmaguest committed:
Suppress compiler warning from boost

We pass the correct type of size_t to raw_function but internally it
then uses it as an int, giving a compiler warning we can do nothing
about

DavidFair committed:
Fix createChildAlg not accepting workspace types

Workspace types were not accepted, this is particularly noticable in Fit
functions. Using the already-existing Workspace registry we can
trivially add a conversion that covers all of these types

DavidFair committed:
Reject duplicated args and kwargs in CreateChildAlg

Rejects a duplicated kwargs and arg, as boost used to handle this before
we took over parsing manually using raw functions.
Now we will explicitly check for this, and throw a invalid_exception (we
couldn't find something for a TypeError though, so there is a potential
future improvement)

gemmaguest committed:
Suppress compiler warning on windows

StephenSmith25 reviewed (commented):
This looks good, comments from #32134 have been addressed. Just one small comment about the cast still. But this is probably good to go. As David said if we need to support more types (e.g. vector of vectors) we can just add them to the typeid checks.

DavidFair commented on scripts/Inelastic/CrystalField/fitting.py:
```diff
@@ -498,6 +498,8 @@ def FWHM(self, value):
                 if self.PhysicalProperty is not None and len(value) == len(self.Temperature) - len(self.PhysicalProperty):
                     value = value + [0] * len(self.PhysicalProperty)
+                    # Cast all types to match the first elem so we don't have mixed lists of int/doubles
+                    value = [type(value[0])(v) for v in value]
```
I was just trying to maintain backwards compat :sweat_smile: since sometimes they passed a list of `int` values so I didn't want to break it

You're completely right about it making no sense as the fitting layer will turn it into a double ultimately

gemmaguest committed:
Always cast FWHM setter to double

Previously we supported passing lists of ints but they will be cast to
double eventually anyway

StephenSmith25 reviewed (approved):
This all looks good, comments here and from #32134 have been address.
